pkg/provider/azure_privatelinkservice.go

```diff
@@ -0,0 +1,682 @@
+	if subnetName == nil {
+		subnetName = &az.SubnetName
+	}
+
+	subnet, existsSubnet, err := az.getSubnet(az.VnetName, *subnetName)
+	if err != nil {
+		return err
+	}
+	if !existsSubnet {
+		return fmt.Errorf("disablePLSNetworkPolicy: failed to get private link service subnet(%s) for service(%s)", *subnetName, serviceName)
+	}
+
+	// Policy already disabled
+	if subnet.PrivateLinkServiceNetworkPolicies == network.VirtualNetworkPrivateLinkServiceNetworkPoliciesDisabled {
+		return nil
+	}
+
+	subnet.PrivateLinkServiceNetworkPolicies = network.VirtualNetworkPrivateLinkServiceNetworkPoliciesDisabled
+	err = az.CreateOrUpdateSubnet(service, subnet)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (az *Cloud) safeDeletePLS(pls *network.PrivateLinkService, service *v1.Service) *retry.Error {
+	if pls == nil {
+		return nil
+	}
+
+	peConns := pls.PrivateEndpointConnections
+	if peConns != nil {
+		for _, peConn := range *peConns {
+			klog.V(2).Infof("deletePLS: deleting PEConnection %s", to.String(peConn.Name))
+			rerr := az.DeletePEConn(service, to.String(pls.Name), to.String(peConn.Name))
+			if rerr != nil {
+				return rerr
+			}
+		}
+	}
+
+	rerr := az.DeletePLS(service, to.String(pls.Name), to.String((*pls.LoadBalancerFrontendIPConfigurations)[0].ID))
+	if rerr != nil {
+		return rerr
+	}
+	klog.V(2).Infof("safeDeletePLS(%s) finished", to.String(pls.Name))
+	return nil
+}
+
+// getPrivateLinkServiceName() returns the name of private link service, or any error
+func (az *Cloud) getPrivateLinkServiceName(
+	existingPLS *network.PrivateLinkService,
+	service *v1.Service,
+	fipConfig *network.FrontendIPConfiguration,
+) (string, error) {
+	existingName := existingPLS.Name
+	serviceName := getServiceName(service)
+
+	if nameFromService, found := service.Annotations[consts.ServiceAnnotationPLSName]; found {
+		nameFromService = strings.TrimSpace(nameFromService)
+		if existingName != nil && !strings.EqualFold(to.String(existingName), nameFromService) {
+			return "", fmt.Errorf(
+				"getPrivateLinkServiceName(%s) failed: cannot change existing private link service name (%s) to (%s)",
+				serviceName,
+				to.String(existingName),
+				nameFromService,
+			)
+		}
+		return nameFromService, nil
+	}
+
+	if existingName != nil {
+		return to.String(existingName), nil
+	}
+
+	// default PLS name: pls-<frontendIPConfigName>
+	return fmt.Sprintf("%s-%s", "pls", *fipConfig.Name), nil
+}
+
+// getExpectedPrivateLinkService builds expected PLS object from service spec
+func (az *Cloud) getExpectedPrivateLinkService(
+	existingPLS *network.PrivateLinkService,
+	plsName *string,
+	clusterName *string,
+	service *v1.Service,
+	fipConfig *network.FrontendIPConfiguration,
+) (dirtyPLS bool, err error) {
+	dirtyPLS = false
+
+	if existingPLS == nil {
+		return false, fmt.Errorf("getExpectedPrivateLinkService: existingPLS is nil (unexpected)")
+	}
+
+	if existingPLS.Name == nil || !strings.EqualFold(*existingPLS.Name, *plsName) {
```

Comment:
if a PLS is in a failed state, could we also set dirtyPLS to trigger a reconciling loop again?

Reply:
Sure, thanks!